src/hooks.ts

```diff
@@ -120,15 +106,65 @@ export const useFeature: UseFeature = (featureKey, options = {}, overrides = {})
       })
       .catch(() => {
         /* The user promise or core client promise rejected. */
-        useFeatureLogger.error(`Error initializing client. The core client or user promise(s) rejected.`);
+        logger.error(`Error initializing client. The core client or user promise(s) rejected.`);
       });
 
-    return () => {
+    return (): void => {
       while (cleanupFns.length) {
         cleanupFns.shift()!();
       }
     };
-  }, [optimizely]);
+  };
+};
+
+/**
+ * A React Hook that retrieves the status of a feature flag and its variables, optionally
+ * auto updating those values based on underlying user or datafile changes.
+ *
+ * Note: The react client can become ready AFTER the timeout period.
+ *       ClientReady and DidTimeout provide signals to handle this scenario.
+ */
+export const useFeature: UseFeature = (featureKey, options = {}, overrides = {}) => {
+  const { isServerSide, optimizely, timeout } = useContext(OptimizelyContext);
+  if (!optimizely) {
+    throw new Error('optimizely prop must be supplied via a parent <OptimizelyProvider>');
+  }
+
+  // Helper function to return the current values for isEnabled and variables.
+  const getCurrentValues = useCallback(
+    () => ({
+      isEnabled: optimizely.isFeatureEnabled(featureKey, overrides.overrideUserId, overrides.overrideAttributes),
+      variables: optimizely.getFeatureVariables(featureKey, overrides.overrideUserId, overrides.overrideAttributes),
+    }),
+    [featureKey, overrides]
+  );
+
+  // Set the initial state immediately serverSide
+  const [data, setData] = useState<UseFeatureState>(() => {
+    if (isServerSide) {
+      return getCurrentValues();
+    }
+    return { isEnabled: false, variables: {} };
+  });
+
+  const [clientReady, setClientReady] = useState(isServerSide ? true : false);
+  const [didTimeout, setDidTimeout] = useState(false);
```

Comment:
To reduce the number of arguments going around, is it possible to collapse all this into one `useState` call returning an object with all these things? Then set different parts of it as needed?

Reply:
I initially thought this would be a hassle since the `useState` set state method doesnt extend by default, making it necessary to return the entire state object.

However, I think you're right that it's worth it here, and it makes the typing simpler too.

Set state calls now look like this:

`setState((state: HookState) => Object.assign({}, state, { didTimeout: true }));`